worldedit-core/src/main/java/com/fastasyncworldedit/core/extent/clipboard/io/FastSchematicReaderV3.java

```diff
@@ -0,0 +1,824 @@
+                        );
+                    }
+                    case CACHE_IDENTIFIER_BLOCK_TILE_ENTITIES -> {
+                        cacheStream.skipNBytes(1); // list child type (TAG_Compound)
+                        this.readEntityContainers(
+                                cacheStream,
+                                cacheNbtIn,
+                                DataFixer.FixTypes.BLOCK_ENTITY,
+                                this.provideTileEntityTransformer(clipboard)
+                        );
+                    }
+                    default -> throw new IOException("invalid cache state - got identifier: 0x" + identifier);
+                }
+            }
+        }
+    }
+
+    /**
+     * Reset the main stream of this clipboard and reads all remaining data that could not be read or fixed yet.
+     * Might need two iterations if the DataVersion tag is after the Blocks tag while the Palette inside the Blocks tag is not
+     * at the first position.
+     *
+     * @param clipboard The clipboard to write into.
+     * @throws IOException on I/O error.
+     */
+    private void readRemainingDataReset(Clipboard clipboard) throws IOException {
+        byte type;
+        String tag;
+        outer:
+        while (!this.remainingTags.isEmpty()) {
+            this.reset();
+            skipHeader(this.dataInputStream);
+            while ((type = dataInputStream.readByte()) != NBTConstants.TYPE_END) {
+                tag = dataInputStream.readUTF();
+                byte b = tag.equals("Blocks") ? CACHE_IDENTIFIER_BLOCK :
+                        tag.equals("Biomes") ? CACHE_IDENTIFIER_BIOMES :
+                                tag.equals("Entities") ? CACHE_IDENTIFIER_ENTITIES :
+                                        CACHE_IDENTIFIER_END;
+                if (!this.remainingTags.remove(b)) {
+                    this.nbtInputStream.readTagPayloadLazy(type, 0);
+                    continue;
+                }
+                switch (tag) {
+                    case "Blocks" -> readBlocks(clipboard);
+                    case "Biomes" -> readBiomes(clipboard);
+                    case "Entities" -> readEntities(clipboard);
+                    default -> this.nbtInputStream.readTagPayloadLazy(type, 0); // Should never happen, but just in case
+                }
+                if (this.remainingTags.isEmpty()) {
+                    break outer;
+                }
+            }
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Requires {@link #read()}, {@link #read(UUID)} or {@link #read(UUID, Function)} to be called before.
+     */
+    @Override
+    public OptionalInt getDataVersion() {
+        return this.dataVersion > -1 ? OptionalInt.of(this.dataVersion) : OptionalInt.empty();
+    }
+
+    private void readBlocks(Clipboard target) throws IOException {
+        this.blockPalette = new BlockState[BlockTypesCache.states.length];
```

Comment:
Ensures that `() -> this.blockPalette.length == 0` is always false as it is always read imemdiately in the `readPalette` call..?

Do we also ever used this "default" value? It seems like an unnecessary memory expense

Reply:
oic, I fucked something up there

Reply:
The check is now correct - it's used so the code can skip the palette initializing (aka assigning the block states to the ordinals from the schematic) if the method is called again (see `#readRemainingDataReset`)

Reply:
But it will still always return true (now)?

Reply:
in which case?

Reply:
Yeah nevermind missed that the calling method was called from two places

Reply:
*Theoretically* the spec does not enforce the palette to start with index `0`:
> It is recommended for maximum data compression that your indices start at zero and skip no values

But highly recommends it. In worst case, if some schematic generators don't do it that way the palette will just be read twice - which is okay for me